client/index.ts

```diff
@@ -0,0 +1,353 @@
+        customerAddress: string,
+        id: string,
+        nonce: number,
+        startBlock: number,
+        endBlock: number,
+        contractAddress: string,
+        data: string,
+        gasLimit: number,
+        challengePeriod: number
+    ): Promise<AppointmentReceipt>;
+    /**
+     * Generates a request object that can be used to request an appointment from
+     * a pisa tower. Also sends the request to the PISA tower to receive an appointment receipt.
+     * @param signer A signing function to create a signature.
+     *   Receives a correctly formatted digest of the appointment, must return a signature
+     *   created by the priv key associated with the customerAddress
+     * @param customerAddress
+     * @param id
+     * @param nonce
+     * @param startBlock
+     * @param endBlock
+     * @param contractAddress
+     * @param data
+     * @param gasLimit
+     * @param challengePeriod
+     * @param eventAddress
+     * @param eventABI
+     * @param eventArgs
+     */
+    generateAndExecuteRequest(
+        signer: (digest: string) => Promise<string>,
+        customerAddress: string,
+        id: string,
+        nonce: number,
+        startBlock: number,
+        endBlock: number,
+        contractAddress: string,
+        data: string,
+        gasLimit: number,
+        challengePeriod: number,
+        eventAddress: string,
+        eventABI: string,
+        eventArgs: string
+    ): Promise<AppointmentReceipt>;
+    public async generateAndExecuteRequest(
+        signer: (digest: string) => Promise<string>,
+        customerAddress: string,
+        id: string,
+        nonce: number,
+        startBlock: number,
+        endBlock: number,
+        contractAddress: string,
+        data: string,
+        gasLimit: number,
+        challengePeriod: number,
+        eventAddress?: string,
+        eventABI?: string,
+        eventArgs?: string
+    ): Promise<AppointmentReceipt> {
+        const request =
+            eventAddress && eventABI && eventArgs
+                ? await this.generateRequest(
+                      signer,
+                      customerAddress,
+                      id,
+                      nonce,
+                      startBlock,
+                      endBlock,
+                      contractAddress,
+                      data,
+                      gasLimit,
+                      challengePeriod,
+                      eventAddress,
+                      eventABI,
+                      eventArgs
+                  )
+                : await this.generateRequest(
+                      signer,
+                      customerAddress,
+                      id,
+                      nonce,
+                      startBlock,
+                      endBlock,
+                      contractAddress,
+                      data,
+                      gasLimit,
+                      challengePeriod
+                  );
+        return await this.executeRequest(request);
```

Comment:
If some but not all three optional params, this will silently ignore the other two without giving an error.
I guess you can just call `generateRequest` with all parameters, instead?

Reply:
There is a check in generateRequest that throws an error. Or do you mean something else?

Reply:
You are doing:

```
function foo(params, A?, B?, C?) {
  if (A && B && C) generateRequest(params, A, B, C);
  else generateRequest(params);
}
```
So if for example A is given, but not B and C, then the condition `A && B && C` would be false and `generateRequest(params)` would be called.
I think you should just call `generateRequest(params, A, B, C)`, so that the parameters are actually checked.

Reply:
I dont mean the check in `generateAndExecuteRequest`, I mean the check **inside** `generateRequest`.

Reply:
I mean that if you call generateAndExecuteRequest with only eventAddress and not eventAbi and eventArgs, then generateAndExecuteRequest would call generateRequest without any of the optional params; so no error would be thrown and the eventAddress parameter would be ignored.
But anyway, since generateRequest has proper checks, you can just do:
```
	        const request = await this.generateRequest(
                      signer,
                      customerAddress,
                      id,
                      nonce,
                      startBlock,
                      endBlock,
                      contractAddress,
                      data,
                      gasLimit,
                      challengePeriod,
                      eventAddress,
                      eventABI,
                      eventArgs
                  );
```
With all arguments.

Reply:
Ah, I see.

The type system wont actually let me do as you've suggested because it wont accept the type `string | undefined` for any event args (the generateRequest func does only doesnt have that overload available).

So I've had to do some weird casting to strings.

Reply:
I see, indeed that's not ideal. I guess I'd prefer to repeat the argument check independentently in both functions, then, rather than forcing the type system.

If you instead hate the code duplication more, you can further simplify like this:
```
    public async generateAndExecuteRequest(
        signer: (digest: string) => Promise<string>,
        customerAddress: string,
        id: string,
        nonce: number,
        startBlock: number,
        endBlock: number,
        contractAddress: string,
        data: string,
        gasLimit: number,
        challengePeriod: number,
        eventAddress?: string,
        eventABI?: string,
        eventArgs?: string
    ): Promise<AppointmentReceipt> {
        const request = await this.generateRequest(
            signer,
            customerAddress,
            id,
            nonce,
            startBlock,
            endBlock,
            contractAddress,
            data,
            gasLimit,
            challengePeriod,
            // we need to cast to string here to allow the compiler to accept possibly undefined values
            // this should only be the case if a caller has passed in undefined as one of the event args
            // in which case we can pass it to generateRequest in below where an error will be thrown.
            eventAddress as string,
            eventABI as string,
            eventArgs as string
        );
        return await this.executeRequest(request);
    }
```

Reply:
Fair enought, updated.